### PR TITLE
Check topology during calls to AvailableData and AvailableIndex in peers bootstrapper

### PR DESCRIPTION
**Pull request by richardartoul**
If we implement checks in AvailableData() and AvailableIndex, we can make it such that the peer bootstrapper is skipped automatically in single-node deployments. This will allow us to make the default bootstrapping configuration:

FS --> Peers --> Commitlog --> Noop-All

And this configuration will work for any cluster size / topology and do "the right thing" without operators having to think about bootstrap configuration.

### CODE REVIEW

File: integration/peers_bootstrap_none_available_test.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/m3db/m3db/topology"
+	"github.com/m3db/m3db/topology/testutil"
+	xlog "github.com/m3db/m3x/log"
+
+	"github.com/stretchr/testify/require"
+)
+
+// TestPeersBootstrapNoneAvailable makes sure that a cluster with the peer bootstrapper enabled can still turn on
+// if none of the nodes peers shard states are available or leaving.
+func TestPeersBootstrapNoneAvailable(t *testing.T) {
```
Comment (robskillington) [inline integration/peers_bootstrap_none_available_test.go]: Nice, did we confirm whether or not a new cluster starts with INITIALIZING shards or AVAILABLE shards?

Comment (robskillington) [reply integration/peers_bootstrap_none_available_test.go]: Damn, think they start as available.  We may have to make a change to m3cluster:

File: storage/bootstrap/bootstrapper/peers/source.go [hunk 2]
```diff
@@ -654,6 +671,90 @@ func (s *peersSource) readBlockMetadataAndIndex(
...
+				// taken ownership of the shard.
+			case shard.Unknown:
+
+				// Success cases - We can bootstrap from this host, which is enough to
+				// mark this shard as bootstrappable.
+			case shard.Leaving:
+				fallthrough
+			case shard.Available:
+				shardPeers.numAvailablePeers++
+				break
```
Comment (robskillington) [inline storage/bootstrap/bootstrapper/peers/source.go]: I don't think you need an explicit break here, should break correctly without falling through unless explicitly falling through.

File: storage/bootstrap/bootstrapper/peers/source.go [hunk 3]
```diff
@@ -677,3 +778,54 @@ func (s *peersSource) markIndexResultErrorAsUnfulfilled(
...
+
+	return topologyState, nil
+}
+
+type topologyState struct {
+	majorityReplicas int
+	shardStates      shardStates
+}
+
+type shardStates map[uint32]map[string]hostShardState
```
Comment (robskillington) [inline storage/bootstrap/bootstrapper/peers/source.go]: very nit: Maybe to make this slightly more readable you could `type hostID string` and `type shardID uint32` so it becomes `map[shardID]map[hostID]hostShardState`?

File: topology/consistency_level.go [hunk 4]
```diff
@@ -307,3 +307,68 @@ const (
 	majority         = "majority"
 	unstrictMajority = "unstrict_majority"
 )
+
+// WriteConsistencyAchieved returns a bool indicating whether or not we've received enough
```
Comment (robskillington) [inline topology/consistency_level.go]: +1 for having these available from the `topology` package.

File: topology/consistency_level.go [hunk 5]
```diff
@@ -307,3 +307,68 @@ const (
...
+		}
+	case ConsistencyLevelMajority:
+		if numSuccess >= majority { // Meets majority
+			return true
+		}
+	case ConsistencyLevelOne:
+		if numSuccess > 0 { // Meets one
+			return true
+		}
+	default:
```
Comment (prateek) [inline topology/consistency_level.go]: unrelated to your change, but mind dropping the default and changing the return false to a panic (same as two functions below)

File: topology/consistency_level.go (hunk 4 above)
Comment (prateek) [reply topology/consistency_level.go]: Yeah good call on this.

File: integration/integration.go [hunk 6]
```diff
@@ -117,10 +118,17 @@ func newMultiAddrAdminClient(
 	clientOpts := adminOpts.
 		SetOrigin(origin).
 		SetInstrumentOptions(instrumentOpts).
-		SetTopologyInitializer(topology.NewStaticInitializer(staticOptions)).
 		SetClusterConnectConsistencyLevel(topology.ConnectConsistencyLevelAny).
 		SetClusterConnectTimeout(time.Second)
 
+	if topologyInitializer != nil {
```
Comment (prateek) [inline integration/integration.go]: nit: can simplify this 
```
if topologyInitialzer == nil {
  topologyInitializer = topology.NewStaticInitializer(staticOptions)
}
clientOpts = clientOpts.SetTopologyInitializer(topologyInitializer)
```

File: storage/bootstrap/bootstrapper/peers/source.go [hunk 7]
```diff
@@ -25,6 +25,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/m3db/m3cluster/shard"
```
Comment (prateek) [inline storage/bootstrap/bootstrapper/peers/source.go]: nit: import order

File: integration/peers_bootstrap_none_available_test.go (hunk 1 above)
Comment (richardartoul) [reply integration/peers_bootstrap_none_available_test.go]: Yeah can probably still land this, and then make that change and update M3DB. The comment in that file seems like its more complicated than just changing the default value, will it affect our client as well?

File: storage/bootstrap/bootstrapper/peers/source.go (hunk 3 above)
Comment (richardartoul) [reply storage/bootstrap/bootstrapper/peers/source.go]: sure

File: integration/integration.go (hunk 6 above)
Comment (richardartoul) [reply integration/integration.go]: good call

File: src/dbnode/generated/proto/index/index.pb.go [hunk 8]
```diff
@@ -59,9 +59,9 @@ type IndexInfo struct {
 	BlockStart   int64          `protobuf:"varint,2,opt,name=blockStart,proto3" json:"blockStart,omitempty"`
 	BlockSize    int64          `protobuf:"varint,3,opt,name=blockSize,proto3" json:"blockSize,omitempty"`
 	FileType     int64          `protobuf:"varint,4,opt,name=fileType,proto3" json:"fileType,omitempty"`
-	SnapshotTime int64          `protobuf:"varint,5,opt,name=snapshotTime,proto3" json:"snapshotTime,omitempty"`
-	Segments     []*SegmentInfo `protobuf:"bytes,6,rep,name=segments" json:"segments,omitempty"`
-	Shards       []uint32       `protobuf:"varint,7,rep,packed,name=shards" json:"shards,omitempty"`
+	Shards       []uint32       `protobuf:"varint,5,rep,packed,name=shards" json:"shards,omitempty"`
```
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/generated/proto/index/index.pb.go]: it actually turns out these files were stale for some reason :/ My P.R is fixing them and putting them in the right order

**Pull request merged**